File: TFEgame.py
import sys
import random
import numpy as np
from msvcrt import getwch


class TFEgame:
    
    def __init__(self, dims:tuple[int]=(4, 4)):
        assert dims[0] == dims[1], 'Dimensions must be square'
        self._dims = dims
        self._board = np.array([[0] * dims[0]] * dims[1])
        self._score = 0
        self._moves = 0
        self._albums = {0:None, 2:'TaylorSwift', 4:'Fearless', 8:'SpeakNow', \
                       16:'Red', 32:'1989', 64:'Reputation', 128:'Lover', \
                        256:'Folklore', 512:'Evermore', 1024:'Midnights', \
                        2048:'END'}
        self._max_tile = 2
        self._gen_starting_tiles()
        self.keys = {'w', 'a', 's', 'd'}
        self.directions = {'w':'up()', 'a':'left()', 's':'down()', 'd':'right()'}

    
    def __repr__(self) -> str:
        """ returns a string representing the current state of the board """
        return '\n'.join([' '.join([str(self._board[i][j]) for j in range(self._dims[1])]) for i in range(self._dims[0])])
    
    def board(self) -> np.array:
        """ returns a copy of the game's board"""
        return self._board.copy()
    
    def dims(self) -> tuple[int]:
        """ returns a tuple of ints representing the dimensions of the board """
        return self._dims
    
    def max_tile(self) -> int:
        """ returns the largest tile on the board """
        return self._max_tile
    
    def _gen_starting_tiles(self) -> None:
        self._board = np.zeros_like(self._board)
        for i in range(2):
            empty_spaces = np.argwhere(self._board == 0)
            r, c = empty_spaces[random.randint(0, len(empty_spaces)-1)]
            self._board[r][c] = 2 if random.random() < 0.9 else 4
        # the max tile will be 4 if it exists, otherwise it will be 2
        if 4 in self._board:
            self._max_tile = 4
        else:
            self._max_tile = 2
    
    def score(self) -> int:
        """ returns the game's score """
        return self._score
    
    def _get_key_input(self) -> str:
        keys = {'w', 'a', 's', 'd'}
        while True:
            key = getwch()
            if key in keys:
                return key
            if key == 'q':
                print('quiting game...')
                sys.exit()
            
    def _do_move(self, m) -> bool:
        """ calls the appropriate method to execute a given move input """                
        def collapse(row:np.array) -> np.array:
            """ inplace collapse of a row or column """
            # move all nonzero elements to the left 
            # ('left' is relative to the direction of the move and is accounted for by the caller)
            nonzeros = list(np.nonzero(row)[0])
            for i in range(len(row)):
                if not nonzeros:
                    break
                index = nonzeros.pop(0)
                row[i] = row[index]
                if i != index:
                    row[index] = 0

            # combine like terms and shift accordingly
            for i in range(len(row)):
                elem = row[i]
                if elem == 0:
                    continue
                if i != len(row)-1 and elem == row[i+1]:
                    elem *= 2
                    row[i+1] = 0
                    self._score += elem
                    self._max_tile = max(elem, self._max_tile)

                row[i] = 0
                first_zero = np.where(row == 0)[0][0]
                row[first_zero] = elem

            return row
    
        def left() -> bool:
            changed = False
            for i, row in enumerate(self._board):
                orig = row.copy()
                self._board[i] = collapse(row)
                if not np.array_equal(orig, self._board[i]):
                    changed = True
            return changed

        def right() -> bool:
            changed = False
            for i, row in enumerate(self._board):
                orig = row.copy()
                self._board[i] = np.flip(collapse(np.flip(row)))
                if not np.array_equal(orig, self._board[i]):
                    changed = True
            return changed
        
        def up() -> bool:
            changed = False
            for i in range(len(self._board)):
                orig = self._board[:,i].copy()
                self._board[:,i] = collapse(self._board[:,i])
                if not np.array_equal(orig, self._board[:,i]):
                    changed = True
            return changed
        
        def down() -> bool:
            changed = False
            for i in range(len(self._board)):
                orig = self._board[:,i].copy()
                self._board[:,i] = np.flip(collapse(np.flip(self._board[:,i])))
                if not np.array_equal(orig, self._board[:,i]):
                    changed = True
            return changed

        return eval(self.directions[m])

    def move(self, key=None) -> str:
        """ gets user input for the next move """
        if key is None:
            key = self._get_key_input()
        changed = self._do_move(key)
        if changed:
            self._add_random()
            self._moves += 1
        return key

    def _add_random(self) -> None:
        if self._is_full():
            return
        empty_spaces = np.argwhere(self._board == 0)
        r, c = empty_spaces[random.randint(0, len(empty_spaces)-1)]
        value = 2 if random.random() < 0.9 else 4
        self._board[r][c] = value

    def _is_full(self) -> bool:
        """ returns a boolean indicating whether the board is full """
        return self._board.all()
    
    def can_move(self) -> bool:
        """ returns a boolean indicating whether the player can move """
        if not self._is_full():
            return True
        def check_adjacent(arr) -> bool:
            """ returns boolean representing two same elements are next to each other """
            for i in range(1, len(arr)):
                if arr[i] == arr[i-1]:
                    return True
            return False
        for i in range(self._dims[0]):  # dims must be square so we can check rows and cols together
            row = self._board[i]
            col = self._board[:,i]
            if check_adjacent(row) or check_adjacent(col):
                return True
        return False
    
    def is_playable(self) -> bool:
        """ returns a boolean indicating whether the game is playable """
        return self.can_move() and not self.won()
    
    def won(self) -> bool:
        """ returns a boolean indicating whether the game has been won, 10/2048 has been reached """
        return np.amax(self._board) == 2048
    
    def lost(self) -> bool:
        """ returns a boolean indicating whether the game as been lost """
        return not self.can_move() and not self.won()
    
    def play(self) -> bool:
        """ the game loop """
        self._gen_starting_tiles()
        while self.is_playable():
            print(f'Score: {self._score} \t Moves: {self._moves}')
            print(self, '\n')
            self.move()
        if self.won():
            print(f'You won with a score of {self._score} in {self._moves} moves!')
            return True
        print(f'You lost with a score of {self._score} in {self._moves} moves.')
        return False

if __name__ == '__main__':
    game = TFEgame()
    game.play()

File: TFEgame.java
import java.util.HashMap;
import java.util.HashSet;
import java.util.Random;
import java.util.Scanner;
// import java.util.Arrays;
import java.util.ArrayList;
import java.util.Collections;

public class TFEgame {

    private int[] dims;
    private ArrayList<ArrayList<Integer>> board;
    private int score;
    private int moves;
    private HashMap<Integer, String> albums;
    private int maxTile;
    
    private Random randomGenerator;
    private Scanner scanner;


    public TFEgame(int[] dims) {
        if (dims.length != 2 || dims[0] != dims[1]) {
            throw new IllegalArgumentException("Dimensions must be square");
        }
        this.dims = dims;
        this.board = new ArrayList<ArrayList<Integer>>();
        for (int i = 0; i < dims[0]; i++) {
            ArrayList<Integer> row = new ArrayList<Integer>();
            for (int j = 0; j < dims[1]; j++) {
                row.add(0);
            }
            board.add(row);
        }
        this.score = 0;
        this.moves = 0;

        this.albums = new HashMap<Integer, String>();
        albums.put(0, "_");
        albums.put(2, "TaylorSwift");
        albums.put(4, "Fearless");
        albums.put(8, "SpeakNow");
        albums.put(16, "Red");
        albums.put(32, "1989");
        albums.put(64, "Reputation");
        albums.put(128, "Lover");
        albums.put(256, "Folklore");
        albums.put(512, "Evermore");
        albums.put(1024, "Midnights");
        albums.put(2048, "END");

        this.maxTile = 2;
        this.randomGenerator = new Random();
        this.scanner = new Scanner(System.in);
        genStartingTiles();
    }

    /*
     * method overloading such that no input dims are required
     */
    public TFEgame() {
        this(new int[] {4, 4});
    }

    /*
     * returns a string representing the current state of the board
     */
    public String toString() {
        String s = "";
        for (int i = 0; i < dims[0]; i++) {
            for (int j = 0; j < dims[1]; j++) {
                s += board.get(i).get(j) + " ";
            }
            s += "\n";
        }
        return s;
    }

    /*
     * returns a deepcopy of the board
     */
    public ArrayList<ArrayList<Integer>> getBoard() {     
        ArrayList<ArrayList<Integer>> newBoard = new ArrayList<ArrayList<Integer>>();
        for (int i = 0; i < dims[0]; i++) {
            ArrayList<Integer> row = new ArrayList<Integer>();
            for (int j = 0; j < dims[1]; j++) {
                row.add(board.get(i).get(j));
            }
            newBoard.add(row);
        }
        return newBoard;
    }

    public int[] getDims() {
        return dims.clone();
    }

    public int getMaxTile() {
        return maxTile;
    }

    public int getScore() {
        return score;
    }

    public int getMoves() {
        return moves;
    }

    /*
     * sets all tiles to 0
     */
    private void setZeros() {
        board = new ArrayList<ArrayList<Integer>>();
        for (int i = 0; i < dims[0]; i++) {
            ArrayList<Integer> row = new ArrayList<Integer>();
            for (int j = 0; j < dims[1]; j++) {
                row.add(0);
            }
            board.add(row);
        }
    } 

    /* 
     * sets two random tiles to 2 or 4
     * 90% chance of 2, 10% chance of 4
     */
    private void genStartingTiles() {
        setZeros();
        // first tile
        int x = randomGenerator.nextInt(dims[0]);
        int y = randomGenerator.nextInt(dims[1]);
        board.get(x).set(y, randomGenerator.nextInt(10) < 9 ? 2 : 4);
        // second tile
        x = randomGenerator.nextInt(dims[0]);
        y = randomGenerator.nextInt(dims[1]);
        while (board.get(x).get(y) != 0) {
            x = randomGenerator.nextInt(dims[0]);
            y = randomGenerator.nextInt(dims[1]);
        }
        board.get(x).set(y, randomGenerator.nextInt(10) < 9 ? 2 : 4);
    }

    /*
     * gets an input key from the user and returns it if it's valid
     */
    private char getKeyInput() {
        HashSet<Character> keys = new HashSet<Character>();
        keys.add('w');
        keys.add('a');
        keys.add('s');
        keys.add('d');
        char key = scanner.next().charAt(0);
        while (!keys.contains(key)) {
            key = scanner.next().charAt(0);
        }
        return key;
    }

    /*
     * returns the row at the given index
     */
    private ArrayList<Integer> getRow(ArrayList<ArrayList<Integer>> board, int index) {
        if (index >= board.size())
            throw new IllegalArgumentException("Row index out of bounds");
        ArrayList<Integer> row = new ArrayList<Integer>();
        for (int i = 0; i < board.size(); i++) {
            row.add(board.get(index).get(i));
        }
        return row;
    }

    /*
     * sets the row at the given index to the given array
     */
    private boolean setRow(ArrayList<ArrayList<Integer>> board, ArrayList<Integer> row, int index) {
        if (row.size() != board.size())
            return false;
        board.set(index, row);
        return true;
    }

    /*
     * returns the column at the given index
     */
    private ArrayList<Integer> getCol(ArrayList<ArrayList<Integer>> board, int index) {
        if (index >= board.get(0).size())
            throw new IllegalArgumentException("Column index out of bounds");
        ArrayList<Integer> column = new ArrayList<Integer>();
        for (int i = 0; i < board.get(0).size(); i++) {
            column.add(board.get(i).get(index));
        }
        return column;
    }

    /*
     * sets the column at the given index to the given array
     */
    private boolean setCol(ArrayList<ArrayList<Integer>> board, ArrayList<Integer> col, int index) {
        if (col.size() != board.get(0).size())
            return false;
        for (int i = 0; i < board.get(0).size(); i++) {
            board.get(i).set(index, col.get(i));
        }
        return true;
    }

    /*
     * collapses the given row by combining like nonzero terms and shifting them to the left
     */
    private ArrayList<Integer> collapse(ArrayList<Integer> row) {
        // the output array
        ArrayList<Integer> collapsed = new ArrayList<Integer>();
        // move all nonzero elements to the left
        int index = 0;
        for (int i = 0; i < row.size(); i++) {
            if (row.get(i) != 0) {
                collapsed.add(row.get(i));
                index++;
            }
        }
        // pad the rest of the array with 0s
        for (int i = index; i < row.size(); i++) {
            collapsed.add(0);
        }
        // combine like terms and shift accordingly
        for (int i = 0; i < collapsed.size(); i++) {
            int elem = collapsed.get(i);
            if (elem == 0)
                continue;
            if (i != collapsed.size()-1 && elem == collapsed.get(i+1)) {
                elem *= 2;
                collapsed.set(i+1, 0);
                score += elem;
                maxTile = Math.max(maxTile, elem);
            }
            collapsed.set(i, 0);
            index = collapsed.indexOf(0);
            collapsed.set(index, elem);
        }
        return collapsed;
    }   

    private boolean left() {
        boolean changed = false;
        for (int i = 0; i < dims[0]; i++) {
            ArrayList<Integer> row = getRow(board, i);
            ArrayList<Integer> collapsed = collapse(row);
            if (!row.equals(collapsed))
                changed = true;
            setRow(board, collapsed, i);
        }
        return changed;
    }

    private boolean right() {
        boolean changed = false;
        for (int i = 0; i < dims[0]; i++) {
            ArrayList<Integer> row = getRow(board, i);
            Collections.reverse(row);
            ArrayList<Integer> collapsed = collapse(row);
            if (!row.equals(collapsed))
                changed = true;
            Collections.reverse(collapsed);
            setRow(board, collapsed, i);
        }
        return changed;
    }

    private boolean up() {
        boolean changed = false;
        for (int i = 0; i < dims[1]; i++) {
            ArrayList<Integer> col = getCol(board, i);
            ArrayList<Integer> collapsed = collapse(col);
            if (!col.equals(collapsed))
                changed = true;
            setCol(board, collapsed, i);
        }
        return changed;
    }

    private boolean down() {
        boolean changed = false;
        for (int i = 0; i < dims[1]; i++) {
            ArrayList<Integer> col = getCol(board, i);
            Collections.reverse(col);
            ArrayList<Integer> collapsed = collapse(col);
            if (!col.equals(collapsed))
                changed = true;
            Collections.reverse(collapsed);
            setCol(board, collapsed, i);
        }
        return changed;
    }

    /*
     * calls the appropriate method based on the given key
     */
    private boolean doMove(char key) {
        switch (key) {
            case 'w':
                return up();
            case 'a':
                return left();
            case 's':
                return down();
            case 'd':
                return right();
        }
        return false;  // the array did not change
    }

    /*
     * gets an input from the user and moves the board accordingly
     */
    public char move() {
        char key = getKeyInput();
        boolean changed = doMove(key);
        if (changed) {
            addRandomTile();
            moves++;
        }
        return key;
    }

    /*
     * method overloading: moves the board according to the given key
     */
    public char move(char key) {
        boolean changed = doMove(key);
        if (changed) {
            addRandomTile();
            moves++;
        }
        return key;
    }

    /*
     * adds a random tile to an empty space on the board
     * 90% chance of adding a 2, 10% chance of adding a 4
     */
    private void addRandomTile() {
        int x = randomGenerator.nextInt(dims[0]);
        int y = randomGenerator.nextInt(dims[1]);
        while (board.get(x).get(y) != 0) {
            x = randomGenerator.nextInt(dims[0]);
            y = randomGenerator.nextInt(dims[1]);
        }
        board.get(x).set(y, randomGenerator.nextInt(10) < 9 ? 2 : 4);
    }

    /*
     * returns true if the board is full
     */
    private boolean isFull() {
        for (int i = 0; i < dims[0]; i++) {
            for (int j = 0; j < dims[1]; j++) {
                if (board.get(i).get(j) == 0)
                    return false;
            }
        }
        return true;
    }

    /*
     * returns true if there are any adjacent tiles that are equal
     */
    private boolean checkAdjacent(ArrayList<Integer> arr) {
        for (int i = 0; i < arr.size()-1; i++) {
            if (arr.get(i) == arr.get(i+1))
                return true;
        }
        return false;
    }

    /*
     * returns true if there are moves that can be made
     */
    public boolean canMove() {
        // check if there are any empty spaces
        if (!isFull())
            return true;
        // check if there are any adjacent tiles that are equal
        // check rows
        for (int i = 0; i < dims[0]; i++) {
            ArrayList<Integer> row = getRow(board, i);
            if (checkAdjacent(row))
                return true;
        }
        // check columns
        for (int i = 0; i < dims[1]; i++) {
            ArrayList<Integer> col = getCol(board, i);
            if (checkAdjacent(col))
                return true;
        }
        return false;
    }

    /*
     * returns true if the player has won the game
     */
    public boolean isWon() {
        for (int i = 0; i < dims[0]; i++) {
            for (int j = 0; j < dims[1]; j++) {
                if (board.get(i).get(j) == 2048)
                    return true;
            }
        }
        return false;
    }

    /*
     * returns true if the player has lost the game (no moves can be made and hasn't won)
     */
    public boolean isLost() {
        return !canMove() && !isWon();
    }

    /*
     * returns true if the game is still playable (can move and hasn't won)
     */
    public boolean isPlayable() {
        return canMove() && !isWon();
    }

    /*
     * plays the game until the player wins or loses
     */
    public boolean play() {
        genStartingTiles();
        while (isPlayable()) {
            System.out.println("Score: " + score + "  Moves: " + moves + "  Max Tile: " + maxTile);
            System.out.println(this);
            move();
        }
        System.out.println("Score: " + score + "  Moves: " + moves + "  Max Tile: " + maxTile);
        System.out.println(this);
        if (isWon()) {
            System.out.println("You won!");
            return true;
        } else {
            System.out.println("You lost!");
            return false;
        }
    }

    public static void main(String[] args) {
        TFEgame game = new TFEgame();
        game.play();
    }
}